core/src/main/java/com/linecorp/armeria/common/stream/AsyncMapStreamMessage.java

```diff
@@ -0,0 +1,179 @@
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        source.subscribe(new AsyncMapSubscriber<>(subscriber, function), executor, options);
+    }
+
+    @Override
+    public void abort() {
+        source.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        requireNonNull(cause, "cause");
+        source.abort(cause);
+    }
+
+    private static final class AsyncMapSubscriber<U> implements Subscriber<Object>, Subscription {
+        private final Subscriber<? super U> downstream;
+        private final Function<Object, CompletableFuture<U>> function;
+
+        @Nullable
+        private volatile Subscription upstream;
+        private volatile boolean canceled;
+
+        AsyncMapSubscriber(Subscriber<? super U> downstream,
+                           Function<Object, CompletableFuture<U>> function) {
+            requireNonNull(downstream, "downstream");
+            requireNonNull(function, "function");
+            this.downstream = downstream;
+            this.function = function;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(Object item) {
+            requireNonNull(item, "item");
+
+            if (canceled) {
+                StreamMessageUtil.closeOrAbort(item);
+                return;
+            }
+
+            try {
+                final CompletableFuture<U> future = function.apply(item);
+                requireNonNull(future, "function.apply() returned null");
+                future.whenComplete((res, cause) -> {
+                    try {
+                        if (cause != null) {
+                            upstream.cancel();
+                            downstream.onError(cause);
+                        } else {
+                            requireNonNull(res, "function.apply()'s future completed with null");
+                            downstream.onNext(res);
+                        }
+                    } catch (Throwable ex) {
+                        StreamMessageUtil.closeOrAbort(item, ex);
+                        upstream.cancel();
+                        onError(ex);
+                    }
+                });
+            } catch (Throwable ex) {
+                StreamMessageUtil.closeOrAbort(item, ex);
+                upstream.cancel();
+                onError(ex);
+            }
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            if (canceled) {
+                return;
+            }
+            canceled = true;
+
+            downstream.onError(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            downstream.onComplete();
```

Comment:
We need to defer the event of `onComplete` until all pending items are published.
For example,
```java
public void onNext(Object item) {
    pendingRequest++;
    ...
}

void onNext0(@Nullable U item, @Nullable Throwable cause) {
     if (--pendingRequest == 0 && isCompleting) {
         downstream.onComplete();
     }
}
public void onComplete() {
    if (pendingRequest > 0) {
        isCompleting = true;
    } else {
        downstream.onComplete();
    }
}
```

Reply:
There were some issues with ordering but I think I have this logic correctly now 👍